plugins/synapticsmst/synapticsmst-device.c

```diff
@@ -504,32 +541,502 @@ synapticsmst_device_get_flash_checksum (SynapticsMSTDevice *device,
+{
+	guint32 esmSize = ESM_CODE_SIZE;
+	guint32 checksum = 0;
+	guint32 flash_checksum = 0;
+	gboolean isESMupdted = FALSE;
+	gboolean  eraseFail = FALSE;
+	guint8 rc = 0xFF;
+	guint32 i = 0;
+	guint32 writeOffset, writeIndex;
+	guint32 write_loops = 0;
+	guint32 retryCount, unitSize;
+	SynapticsMSTDevicePrivate *priv = GET_PRIVATE (device);
+	g_autoptr(SynapticsMSTConnection) connection = NULL;
+	guint8 *ptrESMcode = &payload_data[EEPROM_ESM_OFFSET]; 
+
+	connection = synapticsmst_common_new (priv->fd, priv->layer, priv->rad);
+
+	checksum = 0;
+	for (i = 0; i < esmSize; i++) {
+		checksum += *(payload_data + EEPROM_ESM_OFFSET +i);
+	}
+	flash_checksum = 0;
+	if (synapticsmst_device_get_flash_checksum (device,
+				  esmSize, EEPROM_ESM_OFFSET, &flash_checksum, error)) 
+	{
+		if (checksum != flash_checksum) 
+		{
+			/* erase ESM firmware */	
+			for (i=0; i<4; i++)
+			{				
+				rc = synapticsmst_device_set_flash_sector_erase(device, FLASH_SECTOR_ERASE_64K, i + 4);
+				if (rc)
+				{
+					g_set_error_literal (error,	G_IO_ERROR,	G_IO_ERROR_INVALID_DATA,
+								"can't sector erase flash");
+					eraseFail = TRUE;
+					break;
+				}
+				if (eraseFail){
+					break;
+				}				
+			}
+			/* update ESM firmware */
+			rc = 0;
+			retryCount = 0;
+			unitSize = BLOCK_UNIT;//64
+			writeIndex = 0;	
+			#if 1
+			write_loops = (esmSize / unitSize / 6);
+			//g_debug("WriteEachTime 384byte?");
+			#else		
+			write_loops = (esmSize / unitSize);
+			#endif
+			if (esmSize % unitSize)
+				write_loops++;
+			writeOffset = EEPROM_ESM_OFFSET;
+			do
+			{
+
+				for( i = 0; i < write_loops ; i++ )
+				{
+					rc = synapticsmst_common_rc_set_command (connection,
+							 UPDC_WRITE_TO_EEPROM, unitSize, writeOffset, ptrESMcode + writeIndex);
+					if (rc) {
+					/* repeat once */
+						rc = synapticsmst_common_rc_set_command (connection,
+								 UPDC_WRITE_TO_EEPROM, unitSize, writeOffset, ptrESMcode + writeIndex);
+					}
+					if (rc)	{
+						g_set_error_literal (error,	G_IO_ERROR,	G_IO_ERROR_INVALID_DATA,
+									"ESM write fail");
+						break;
+					}
+					writeOffset += unitSize;
+					writeIndex += unitSize;
+				}				
+				/* check ESM checksum */
+				checksum = 0;
+				for (i = 0; i < esmSize; i++) {
+					checksum += *(payload_data + EEPROM_ESM_OFFSET +i);
+				}
+				rc = synapticsmst_device_get_flash_checksum (device,
+					esmSize, EEPROM_ESM_OFFSET, &flash_checksum, error);
+				if(checksum == flash_checksum){
+					isESMupdted = TRUE;
+					break;
+				}
+			}while(retryCount++ < MAX_RETRY_COUNTS);
+
+			if(rc && retryCount== MAX_RETRY_COUNTS){
+				g_set_error_literal (error,	G_IO_ERROR,	G_IO_ERROR_INVALID_DATA,
+									"ESM update fail");
+			}
+			if(isESMupdted == TRUE){				
+				return 1;/* ESM update done */	
+			}
+			else{				
+				return 0;/* ESM update fail */
+			}
+		}
+		else{			
+			return 2;/* ESM checksum same */
+		}
+	}
+	else {
+		return 0;/* Get chcksum fail */
+
```

Comment:
Rather than specific error codes meaning different things, can you please instead make the return type `gboolean`?  Then you return `TRUE` for checksum same or update done, and `FALSE` for checksum fail.  If checksum failed then set the `GError` so it can be propogated (and other functions might prefix it).